# Moises without the Noises — Song Processing Notebook

Runs on Colab's free GPU to produce everything the app needs for one song:
- **6-stem separation** via Demucs `htdemucs_6s` (vocals, drums, bass, other, guitar, piano)
- **Lyric transcription** via OpenAI Whisper (word-level timestamps)
- **BPM + beat detection** via librosa
- **Key detection** via librosa (Krumhansl-Schmuckler)
- **Note detection** via librosa pyin (vocals, bass)

Output is a zip you drop directly into `backend/data/`. The app picks it up automatically.

---

### Before running
**Runtime → Change runtime type → T4 GPU** (free tier is fine)

### Mobile data warning
The first run downloads ~3.5 GB of model weights (Demucs + Whisper medium).
These are cached in `/root/.cache/` for the duration of the Colab session,
but **are lost when the session ends** unless you mount Google Drive (Cell 2 shows how).
Subsequent cells in the same session cost ~0 extra data.

### Known limitations
- Note detection uses `librosa.pyin` — a **monophonic** pitch tracker. Reliable on
  isolated bass and vocal stems. The `other` stem is often polyphonic and results there
  are best-effort. Guitar and piano stems don't get note detection (chords).
- Whisper transcription is English-optimized by default. Change `WHISPER_LANGUAGE`
  below for other languages.
- `htdemucs_6s` is slightly slower than `htdemucs` (4-stem). On a T4, expect
  ~30–90 seconds per song depending on length.

## Cell 1 — Install dependencies

This installs **Demucs** by default.
If you selected a different engine in Cell 3, install its package here too.

| Engine choice | Extra install command |
|---|---|
| `htdemucs_ft` / `htdemucs_6s` | *(already installed below)* |
| `spleeter:*` | `!pip install -q spleeter` |
| `umxl` / `umxhq` | `!pip install -q openunmix torchaudio` |
| `mdx-vocalft` / `mdx-inst-hq3` | `!pip install -q 'audio-separator[cpu]'` |
| `bs-roformer` | `!pip install -q bs-roformer-infer` |
| `melband-roformer` | `!pip install -q melband-roformer-infer` |

**First-run data cost** (model weights, cached for the Colab session):
- Demucs htdemucs_ft / htdemucs_6s: ~2.0–2.3 GB
- Spleeter: ~1.0–1.2 GB (includes TensorFlow)
- Open-Unmix umxl / umxhq: ~0.4 GB
- MDX-Net mdx-vocalft / mdx-inst-hq3: ~0.35 GB per model
- BS-RoFormer: ~0.4 GB
- Mel-Band RoFormer: ~0.3 GB

⚠️ **Mobile data warning:** run these installs on Wi-Fi the first time.

In [1]:
# Base install — always needed
!pip install -q demucs openai-whisper librosa soundfile

# ── Optional engines — uncomment the one you need ─────────────────────────
# Spleeter (installs TensorFlow ~500 MB — only if you chose spleeter:*)
# !pip install -q spleeter

# Open-Unmix
# !pip install -q openunmix torchaudio

# MDX-Net via audio-separator
# !pip install -q 'audio-separator[cpu]'

# BS-RoFormer
# !pip install -q bs-roformer-infer

# Mel-Band RoFormer
# !pip install -q melband-roformer-infer

# Verify GPU is available
import torch
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f'Device: {device}')
if device == 'cpu':
    print('WARNING: No GPU detected. Processing will be significantly slower.')
    print('Go to Runtime → Change runtime type → GPU and re-run.')
else:
    print(f'GPU: {torch.cuda.get_device_name(0)}')


Device: cuda
GPU: Tesla T4


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## Cell 2 — (Optional) Mount Google Drive to cache model weights

Skip this if you're on an unmetered connection and don't mind re-downloading
weights each session (~3.5 GB). If you're on mobile data, run this once to
cache weights to your Drive and avoid the download on future sessions.

In [3]:
CACHE_WEIGHTS_TO_DRIVE = False  # Set to True to enable Drive caching

if CACHE_WEIGHTS_TO_DRIVE:
    from google.colab import drive
    drive.mount('/content/drive')

    import os
    DRIVE_CACHE = '/content/drive/MyDrive/mwtn_model_cache'
    os.makedirs(DRIVE_CACHE, exist_ok=True)

    # Point both Demucs and Whisper at the Drive cache directory
    # so weights downloaded once persist across sessions.
    os.environ['TORCH_HOME'] = DRIVE_CACHE
    os.environ['XDG_CACHE_HOME'] = DRIVE_CACHE
    print(f'Model weights will be cached to: {DRIVE_CACHE}')
    print('First run still downloads weights; subsequent sessions will load from Drive.')
else:
    print('Drive caching disabled. Weights download fresh each session.')

Drive caching disabled. Weights download fresh each session.


## Cell 3 — Configuration

Set your preferences here before running the rest.

In [4]:
# ════════════════════════════════════════════════════════════════════════
# CONFIGURATION — edit this cell, then run all cells top to bottom
# ════════════════════════════════════════════════════════════════════════

# Backward compatibility for older Colab configs that still set MODEL
# instead of SEPARATION_MODEL / SEPARATION_ENGINE.
if 'MODEL' in globals():
    MODEL = str(MODEL)
else:
    MODEL = 'htdemucs_6s'

SEPARATION_MODEL = globals().get('SEPARATION_MODEL', MODEL)
SEPARATION_MODEL = str(SEPARATION_MODEL or MODEL)

if 'WHISPER_MODEL' not in globals() or not globals().get('WHISPER_MODEL'):
    WHISPER_MODEL = 'medium'
if 'WHISPER_LANGUAGE' not in globals():
    WHISPER_LANGUAGE = None

ENGINE_DISPATCH = {
    'htdemucs_ft': 'demucs', 'htdemucs_6s': 'demucs',
    'spleeter:2stems': 'spleeter', 'spleeter:4stems': 'spleeter', 'spleeter:5stems': 'spleeter',
    'umxl': 'openunmix', 'umxhq': 'openunmix',
    'mdx-vocalft': 'mdxnet', 'mdx-inst-hq3': 'mdxnet',
    'bs-roformer': 'bs_roformer',
    'melband-roformer': 'melband_roformer',
}

if 'SEPARATION_ENGINE' not in globals() or not globals().get('SEPARATION_ENGINE'):
    SEPARATION_ENGINE = ENGINE_DISPATCH.get(SEPARATION_MODEL, 'demucs')

# ── Choose your separation engine ────────────────────────────────────────
#
# ┌─────────────────────────┬────────┬────────────────────────────────────────┐
# │ SEPARATION_MODEL        │ Stems  │ Best for                               │
# ├─────────────────────────┼────────┼────────────────────────────────────────┤
# │ htdemucs_6s  (default)  │ 6      │ Multi-instrument practice (guitar+piano)│
# │ htdemucs_ft             │ 4      │ Fast, clean — no guitar/piano needed   │
# ├─────────────────────────┼────────┼────────────────────────────────────────┤
# │ spleeter:2stems         │ 2      │ Karaoke only (vocals + accompaniment)  │
# │ spleeter:4stems         │ 4      │ Faster on GPU than Demucs              │
# │ spleeter:5stems         │ 5      │ + piano split                          │
# ├─────────────────────────┼────────┼────────────────────────────────────────┤
# │ umxl                    │ 4      │ Research baseline, stable, reproducible │
# │ umxhq                   │ 4      │ MUSDB18-HQ baseline                    │
# ├─────────────────────────┼────────┼────────────────────────────────────────┤
# │ mdx-vocalft             │ 2      │ Clean vocals on reverb-heavy material  │
# │ mdx-inst-hq3            │ 2      │ Best instrumental extraction (2-stem)  │
# ├─────────────────────────┼────────┼────────────────────────────────────────┤
# │ bs-roformer             │ 6      │ Best overall quality (state of the art)│
# │ melband-roformer        │ 2      │ Best vocal isolation specifically       │
# └─────────────────────────┴────────┴────────────────────────────────────────┘
#
# ⚠️  If you choose spleeter:*, bs-roformer, melband-roformer, umxl/umxhq,
#    or mdx-*, uncomment the matching install line in Cell 1 and re-run it.

# If older notebook code still uses MODEL, it gets mapped here.
if 'MODEL' in globals() and MODEL:
    SEPARATION_MODEL = str(MODEL)
else:
    SEPARATION_MODEL = 'htdemucs_6s'

# ── Whisper transcription settings ───────────────────────────────────────
#   tiny   ~75 MB  — fastest, least accurate
#   base   ~150 MB — fast, decent
#   small  ~500 MB — good balance
#   medium ~1.5 GB — best open-source accuracy (recommended on GPU)
#   large  ~3 GB   — best, slow even on GPU
WHISPER_MODEL = 'medium'

# None = auto-detect. Explicit is faster and more accurate when you know it.
# Options: 'english', 'spanish', 'portuguese', 'français', 'italian', etc.
WHISPER_LANGUAGE = None

# ─────────────────────────────────────────────────────────────────────────
# Derive which engine handles this model (used by Cell 5)
MDX_MODEL_KEYS = {
    'mdx-vocalft': 'UVR_MDXNET_KARA_2',
    'mdx-inst-hq3': 'UVR-MDX-NET-Inst_HQ_3',
}
SEPARATION_ENGINE = ENGINE_DISPATCH.get(SEPARATION_MODEL, 'demucs')

print(f'Separation model:  {SEPARATION_MODEL}')
print(f'Separation engine: {SEPARATION_ENGINE}')
print(f'Whisper model:     {WHISPER_MODEL}')
print(f'Language:          {WHISPER_LANGUAGE or "auto-detect"}')

# ── Google Drive output folder ────────────────────────────────────────────
# This MUST match the "mwtn output folder name" field in the desktop app's
# Settings panel (gear icon). The default in both places is 'mwtn-outputs'.
# If you change it here, change it there too — that is the gap that causes
# "songs not found" in the desktop app.
if 'DRIVE_OUTPUT_FOLDER' not in globals() or not globals().get('DRIVE_OUTPUT_FOLDER'):
    DRIVE_OUTPUT_FOLDER = 'mwtn-outputs'

from pathlib import Path
DRIVE_OUTPUT_DIR = Path('/content/drive/MyDrive') / DRIVE_OUTPUT_FOLDER
DRIVE_OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

print(f'Drive output folder: My Drive/{DRIVE_OUTPUT_FOLDER}/')
print(f'Drive output dir: {DRIVE_OUTPUT_DIR}')


Demucs model:  htdemucs_6s
Whisper model: medium
Language:      auto-detect


## Cell 4 — Load your song

**Pick ONE of the three methods below** and set `UPLOAD_METHOD` accordingly.
Comment out the other two blocks. Any common audio format works: mp3, wav, flac, m4a, ogg.

| Method | When to use |
|---|---|
| `'drive'` | **Recommended.** Upload the file to Google Drive first, then paste its path. Most reliable. |
| `'url'` | Song is publicly accessible via a direct download URL (e.g. a Dropbox/GDrive share link). |
| `'widget'` | Last resort. The Colab file-picker widget — works but can time out if you're slow to click. |

**Data cost:** just the song file (~3–10 MB for mp3).

In [5]:
from pathlib import Path

# ── Pick your method ──────────────────────────────────────────────────────────
UPLOAD_METHOD = 'drive'  # 'drive' | 'url' | 'widget'

# Used by 'drive' method: path relative to /content/drive/MyDrive/
# Example: 'Music/bohemian_rhapsody.mp3'
DRIVE_FILE_PATH = 'Music/Dunsin-Oyekan-You-Remain-Thesame-1.mp3'

# Used by 'url' method: must be a DIRECT download link, not a webpage.
# Dropbox: change ?dl=0 to ?dl=1 at the end of the share URL.
# Google Drive: use https://drive.google.com/uc?export=download&id=FILE_ID
DIRECT_URL = ''
# ─────────────────────────────────────────────────────────────────────────────

input_filename = None

if UPLOAD_METHOD == 'drive':
    from google.colab import drive
    drive.mount('/content/drive', force_remount=False)
    src = Path('/content/drive/MyDrive') / DRIVE_FILE_PATH
    if not src.exists():
        raise FileNotFoundError(
            f'File not found in Drive: {src}\n'
            f'Make sure the file is uploaded to your Google Drive at that path.'
        )
    # Copy to Colab local storage so Demucs can write next to it freely.
    import shutil
    dest = Path('/content') / src.name
    shutil.copy(src, dest)
    input_filename = str(dest)
    print(f'Loaded from Drive: {src.name}')

elif UPLOAD_METHOD == 'url':
    import urllib.request
    if not DIRECT_URL:
        raise ValueError('Set DIRECT_URL to a direct download link.')
    # Derive filename from URL; fall back to 'song.mp3' if URL has no extension.
    url_filename = Path(DIRECT_URL.split('?')[0]).name or 'song.mp3'
    dest = Path('/content') / url_filename
    print(f'Downloading {url_filename}...')
    urllib.request.urlretrieve(DIRECT_URL, dest)
    input_filename = str(dest)
    print(f'Downloaded: {dest} ({dest.stat().st_size / 1e6:.1f} MB)')

elif UPLOAD_METHOD == 'widget':
    # The original file-picker widget. Works, but blocks until you select
    # a file — if you take too long or the popup is blocked, it raises
    # KeyboardInterrupt. Click the 'Choose Files' button immediately.
    from google.colab import files
    print('A file picker will appear below. Select your file immediately.')
    uploaded = files.upload()
    if not uploaded:
        raise RuntimeError('No file selected.')
    input_filename = list(uploaded.keys())[0]
    print(f'Uploaded: {input_filename}')

else:
    raise ValueError(f'Unknown UPLOAD_METHOD: {UPLOAD_METHOD!r}. Use drive, url, or widget.')

input_path = Path(input_filename)
print(f'Song file: {input_path} ({input_path.stat().st_size / 1e6:.1f} MB)')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Loaded from Drive: Dunsin-Oyekan-You-Remain-Thesame-1.mp3
Song file: /content/Dunsin-Oyekan-You-Remain-Thesame-1.mp3 (13.7 MB)


## Cell 5 — Stem separation

Dispatches to the engine you chose in Cell 3.

Typical GPU runtimes for a 4-min song:
- Demucs htdemucs_ft / htdemucs_6s: 30–90 s
- Spleeter: 10–30 s (TF, very fast on GPU)
- Open-Unmix umxl: 20–60 s
- MDX-Net: 15–45 s (ONNX Runtime)
- BS-RoFormer: 30–90 s
- Mel-Band RoFormer: 20–60 s

⚠️ CPU is 5–30× slower than GPU for all engines.

In [6]:
import subprocess
import sys
import shutil
from pathlib import Path

# input_path is set by Cell 4.
import re
raw_stem = input_path.stem
song_id = re.sub(r'[^\w]', '_', raw_stem).strip('_') or 'song'
print(f'Song ID: {song_id}')
print(f'Running engine: {SEPARATION_ENGINE} (model: {SEPARATION_MODEL})')

# stems_source_dir is set by the chosen engine block below.
stems_source_dir = None

# ── Demucs ────────────────────────────────────────────────────────────────
if SEPARATION_ENGINE == 'demucs':
    demucs_out = Path('_demucs_raw')
    result = subprocess.run(
        [sys.executable, '-m', 'demucs', '-n', SEPARATION_MODEL,
         '-o', str(demucs_out), str(input_path)],
        capture_output=False, text=True,
    )
    if result.returncode != 0:
        raise RuntimeError('Demucs failed — check output above.')
    stems_source_dir = demucs_out / SEPARATION_MODEL / input_path.stem

# ── Spleeter ──────────────────────────────────────────────────────────────
elif SEPARATION_ENGINE == 'spleeter':
    spleeter_out = Path('_spleeter_raw')
    result = subprocess.run(
        [sys.executable, '-m', 'spleeter', 'separate',
         '-p', SEPARATION_MODEL, '-o', str(spleeter_out), str(input_path)],
        capture_output=False, text=True,
    )
    if result.returncode != 0:
        raise RuntimeError('Spleeter failed — check output above.')
    stems_source_dir = spleeter_out / input_path.stem

# ── Open-Unmix ────────────────────────────────────────────────────────────
elif SEPARATION_ENGINE == 'openunmix':
    import torch, torchaudio, soundfile as sf, numpy as np
    print(f'Loading Open-Unmix ({SEPARATION_MODEL})…')
    separator = torch.hub.load('sigsep/open-unmix-pytorch', SEPARATION_MODEL, trust_repo=True)
    separator.eval()
    audio, sr = torchaudio.load(str(input_path))
    if audio.shape[0] == 1:
        audio = audio.repeat(2, 1)
    audio = audio.unsqueeze(0)
    if sr != separator.sample_rate:
        audio = torchaudio.transforms.Resample(sr, separator.sample_rate)(audio)
    print('Separating…')
    with torch.no_grad():
        estimates = separator(audio)
    umx_out = Path('_umx_raw') / song_id
    umx_out.mkdir(parents=True, exist_ok=True)
    for i, name in enumerate(['vocals', 'drums', 'bass', 'other']):
        wav = estimates[0, i].cpu().numpy().T
        sf.write(str(umx_out / f'{name}.wav'), wav, separator.sample_rate)
        print(f'  wrote {name}.wav')
    stems_source_dir = umx_out

# ── MDX-Net ───────────────────────────────────────────────────────────────
elif SEPARATION_ENGINE == 'mdxnet':
    from audio_separator.separator import Separator
    mdx_model_key = MDX_MODEL_KEYS[SEPARATION_MODEL]
    mdx_out = Path('_mdx_raw')
    mdx_out.mkdir(exist_ok=True)
    print(f'Running MDX-Net ({mdx_model_key})…')
    sep = Separator(model_name=mdx_model_key, output_dir=str(mdx_out), output_format='WAV')
    sep.load_model()
    sep.separate(str(input_path))
    # Rename outputs to vocals.wav / instrumental.wav
    renamed_out = Path('_mdx_renamed') / song_id
    renamed_out.mkdir(parents=True, exist_ok=True)
    for f in mdx_out.glob('*.wav'):
        nl = f.stem.lower()
        if 'vocal' in nl:
            shutil.copy(f, renamed_out / 'vocals.wav')
        elif 'instrumental' in nl or 'accompaniment' in nl:
            shutil.copy(f, renamed_out / 'instrumental.wav')
        else:
            shutil.copy(f, renamed_out / f.name)
    stems_source_dir = renamed_out

# ── BS-RoFormer ───────────────────────────────────────────────────────────
elif SEPARATION_ENGINE == 'bs_roformer':
    from bs_roformer import BSRoformerSession
    bsr_out = Path('_bsroformer_raw')
    bsr_out.mkdir(exist_ok=True)
    print('Running BS-RoFormer (downloads ~400 MB on first use)…')
    with BSRoformerSession() as session:
        session.infer(str(input_path.parent), store_dir=str(bsr_out))
    # Normalise stem names
    norm_out = Path('_bsroformer_norm') / song_id
    norm_out.mkdir(parents=True, exist_ok=True)
    expected = ['vocals', 'drums', 'bass', 'guitar', 'piano', 'other', 'instrumental']
    for wav in bsr_out.glob('*.wav'):
        for stem_name in expected:
            if f'_{stem_name}.' in wav.name.lower():
                shutil.copy(wav, norm_out / f'{stem_name}.wav')
                break
        else:
            shutil.copy(wav, norm_out / wav.name)
    stems_source_dir = norm_out

# ── Mel-Band RoFormer ─────────────────────────────────────────────────────
elif SEPARATION_ENGINE == 'melband_roformer':
    from mel_band_roformer import MelBandRoformerSession
    mb_out = Path('_melband_raw')
    mb_out.mkdir(exist_ok=True)
    print('Running Mel-Band RoFormer (downloads ~300 MB on first use)…')
    with MelBandRoformerSession() as session:
        session.infer(str(input_path.parent), store_dir=str(mb_out))
    norm_out = Path('_melband_norm') / song_id
    norm_out.mkdir(parents=True, exist_ok=True)
    for wav in mb_out.glob('*.wav'):
        nl = wav.stem.lower()
        if 'vocal' in nl:
            shutil.copy(wav, norm_out / 'vocals.wav')
        elif 'instrumental' in nl or 'accompaniment' in nl:
            shutil.copy(wav, norm_out / 'instrumental.wav')
        else:
            shutil.copy(wav, norm_out / wav.name)
    stems_source_dir = norm_out

else:
    raise ValueError(f'Unknown SEPARATION_ENGINE: {SEPARATION_ENGINE}')

stem_wavs = list(stems_source_dir.glob('*.wav'))
print(f'\nSeparation complete. Stems: {[w.stem for w in stem_wavs]}')


Song ID: Dunsin_Oyekan_You_Remain_Thesame_1
Running Demucs (htdemucs_6s)...

Separation complete. Stems: ['piano', 'guitar', 'bass', 'drums', 'vocals', 'other']


## Cell 6 — Lyric transcription (Whisper)

Transcribes the **vocals stem** (not the full mix) for cleaner results.
Returns word-level timestamps the frontend uses to sync lyrics to playback.

This cell is skipped automatically if no vocals stem is available.

In [7]:
import whisper
import json

vocals_path = stems_source_dir / 'vocals.wav'

lyrics_result = None

if not vocals_path.exists():
    print('No vocals stem found — skipping transcription.')
else:
    print(f'Loading Whisper ({WHISPER_MODEL})...')
    whisper_model = whisper.load_model(WHISPER_MODEL)

    print('Transcribing vocals stem...')
    transcribe_kwargs = {
        'word_timestamps': True,  # needed for word-level sync in the frontend
        'verbose': False,
    }
    if WHISPER_LANGUAGE:
        transcribe_kwargs['language'] = WHISPER_LANGUAGE

    result = whisper_model.transcribe(str(vocals_path), **transcribe_kwargs)

    # Flatten to word-level segments for the frontend.
    # Each word: {start, end, word}
    words = []
    for seg in result.get('segments', []):
        for w in seg.get('words', []):
            words.append({
                'start': round(w['start'], 3),
                'end': round(w['end'], 3),
                'word': w['word'].strip(),
            })

    # Also keep full-segment text for display
    segments = [{
        'start': round(s['start'], 3),
        'end': round(s['end'], 3),
        'text': s['text'].strip(),
    } for s in result.get('segments', [])]

    lyrics_result = {
        'language': result.get('language', 'unknown'),
        'words': words,
        'segments': segments,
    }

    print(f'Language detected: {lyrics_result["language"]}')
    print(f'Words transcribed: {len(words)}')
    if segments:
        print(f'First line: {segments[0]["text"]}')

    # Free GPU memory — Whisper is no longer needed
    del whisper_model
    torch.cuda.empty_cache()

Loading Whisper (medium)...


100%|██████████████████████████████████████| 1.42G/1.42G [00:12<00:00, 120MiB/s]


Transcribing vocals stem...
Detected language: Yoruba


100%|██████████| 56907/56907 [01:06<00:00, 849.45frames/s] 

Language detected: yo
Words transcribed: 763
First line: A


## Cell 7 — BPM & beat detection

In [8]:
import librosa
import numpy as np

# Use drums stem for beat detection if available — most reliable signal.
# Fall back to 'other' or whatever is available.
beat_stem_priority = ['drums', 'other', 'bass', 'vocals']
beat_stem_path = None
for stem_name in beat_stem_priority:
    candidate = stems_source_dir / f'{stem_name}.wav'
    if candidate.exists():
        beat_stem_path = candidate
        print(f'Using {stem_name} stem for BPM detection')
        break

beats_result = None

if beat_stem_path:
    y, sr = librosa.load(str(beat_stem_path), sr=None, mono=True)
    tempo, beat_frames = librosa.beat.beat_track(y=y, sr=sr, units='frames')
    beat_times = librosa.frames_to_time(beat_frames, sr=sr).tolist()
    downbeats = beat_times[::4]  # Estimate: every 4th beat

    beats_result = {
        'bpm': round(float(np.squeeze(tempo)), 2),
        'beats': [round(t, 4) for t in beat_times],
        'downbeats': [round(t, 4) for t in downbeats],
    }
    print(f'BPM: {beats_result["bpm"]}')
    print(f'Beat count: {len(beat_times)}')
else:
    print('No stem available for BPM detection — skipping.')

Using drums stem for BPM detection


## Cell 8 — Key detection

In [ ]:
# Use 'other' stem for key — it has the richest harmonic content.
key_stem_priority = ['other', 'vocals', 'guitar', 'piano', 'bass']
key_stem_path = None
for stem_name in key_stem_priority:
    candidate = stems_source_dir / f'{stem_name}.wav'
    if candidate.exists():
        key_stem_path = candidate
        print(f'Using {stem_name} stem for key detection')
        break

key_result = None

if key_stem_path:
    y, sr = librosa.load(str(key_stem_path), sr=None, mono=True)

    # HPSS: separate harmonic from percussive before chroma analysis.
    # Percussion smears chroma — this improves key detection accuracy.
    y_harmonic, _ = librosa.effects.hpss(y)
    chroma = librosa.feature.chroma_cqt(y=y_harmonic, sr=sr)
    chroma_mean = chroma.mean(axis=1)

    # Krumhansl-Schmuckler key profiles
    major_template = np.array([6.35, 2.23, 3.48, 2.33, 4.38, 4.09,
                                2.52, 5.19, 2.39, 3.66, 2.29, 2.88])
    minor_template = np.array([6.33, 2.68, 3.52, 5.38, 2.60, 3.53,
                                2.54, 4.75, 3.98, 2.69, 3.34, 3.17])
    note_names = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B']

    best_key, best_mode, best_corr = None, None, -np.inf
    for i in range(12):
        for template, mode in [(major_template, 'major'), (minor_template, 'minor')]:
            corr = np.corrcoef(chroma_mean, np.roll(template, i))[0, 1]
            if corr > best_corr:
                best_corr, best_key, best_mode = corr, i, mode

    root = note_names[best_key]
    key_result = {
        'key': f'{root} {best_mode}',
        'root': root,
        'mode': best_mode,
        'confidence': round(float(best_corr), 3),
    }
    print(f'Key: {key_result["key"]} (confidence: {key_result["confidence"]})')
else:
    print('No stem available for key detection — skipping.')

Using other stem for key detection


## Cell 9 — Note detection (vocals + bass)

Uses `librosa.pyin` — a monophonic pitch tracker. Reliable on isolated
vocals and bass stems. Guitar/piano are polyphonic and are skipped here.

In [ ]:
import json

FREQ_RANGES = {
    'bass':   ('C1', 'G4'),
    'vocals': ('C2', 'C6'),
}
MIN_SEGMENT_DURATION = 0.08
NOTE_CHANGE_THRESHOLD_SEMITONES = 0.5


def extract_note_timeline(audio_path, instrument):
    """
    Identical to backend/note_extraction.py — duplicated here since Colab
    can't import from your local project. If you change the logic in one
    place, update the other to match.
    """
    fmin = librosa.note_to_hz(FREQ_RANGES[instrument][0])
    fmax = librosa.note_to_hz(FREQ_RANGES[instrument][1])

    y, sr = librosa.load(audio_path, sr=None, mono=True)
    f0, voiced_flag, _ = librosa.pyin(y, fmin=fmin, fmax=fmax, sr=sr)
    times = librosa.times_like(f0, sr=sr)

    raw_points = []
    for t, f, voiced in zip(times, f0, voiced_flag):
        if voiced and f is not None and not np.isnan(f):
            raw_points.append((float(t), librosa.hz_to_midi(float(f))))

    if not raw_points:
        return []

    frame_hop = times[1] - times[0] if len(times) > 1 else 0.01
    gap_threshold = frame_hop * 3

    segments = []
    seg_start = raw_points[0][0]
    seg_midi = [raw_points[0][1]]
    prev_t = raw_points[0][0]

    for t, midi in raw_points[1:]:
        avg = sum(seg_midi) / len(seg_midi)
        if abs(midi - avg) <= NOTE_CHANGE_THRESHOLD_SEMITONES and (t - prev_t) <= gap_threshold:
            seg_midi.append(midi)
        else:
            segments.append((seg_start, prev_t, seg_midi))
            seg_start, seg_midi = t, [midi]
        prev_t = t
    segments.append((seg_start, prev_t, seg_midi))

    cleaned = []
    for start, end, midi_values in segments:
        avg_midi = round(sum(midi_values) / len(midi_values))
        note_name = librosa.midi_to_note(avg_midi)
        duration = end - start
        if cleaned:
            ps, pe, pn, pm = cleaned[-1]
            if duration < MIN_SEGMENT_DURATION or (note_name == pn and (start - pe) <= gap_threshold):
                cleaned[-1] = (ps, end, pn, pm)
                continue
        cleaned.append((start, end, note_name, avg_midi))

    return [{'start': round(s, 3), 'end': round(e, 3), 'note': n, 'midi': m} for s, e, n, m in cleaned]


notes_by_stem = {}
for instrument in FREQ_RANGES:
    stem_path = stems_source_dir / f'{instrument}.wav'
    if not stem_path.exists():
        print(f'{instrument}: stem not found, skipping note detection')
        continue
    print(f'Extracting notes for {instrument}...')
    timeline = extract_note_timeline(str(stem_path), instrument)
    notes_by_stem[instrument] = timeline
    print(f'  {len(timeline)} segments found')

print('Note detection complete.')

## Cell 10 — Assemble output & download

Packages everything into a zip. Extract it so `<song_id>/manifest.json`
lands directly inside your local `backend/data/` — not nested inside
an extra folder.

**Data cost:** the zip is mostly the WAV stem files.
Stems are typically 20–60 MB each × 6 stems = 120–360 MB total.
Download this on Wi-Fi.

In [ ]:
i
m
p
o
r
t
 
s
h
u
t
i
l

i
m
p
o
r
t
 
j
s
o
n

f
r
o
m
 
p
a
t
h
l
i
b
 
i
m
p
o
r
t
 
P
a
t
h


o
u
t
p
u
t
_
d
i
r
 
=
 
P
a
t
h
(
'
o
u
t
p
u
t
'
)
 
/
 
s
o
n
g
_
i
d

o
u
t
p
u
t
_
d
i
r
.
m
k
d
i
r
(
p
a
r
e
n
t
s
=
T
r
u
e
,
 
e
x
i
s
t
_
o
k
=
T
r
u
e
)


#
 
C
o
p
y
 
s
t
e
m
s

s
t
e
m
_
f
i
l
e
s
 
=
 
{
}

f
o
r
 
w
a
v
 
i
n
 
s
t
e
m
s
_
s
o
u
r
c
e
_
d
i
r
.
g
l
o
b
(
'
*
.
w
a
v
'
)
:

 
 
 
 
d
e
s
t
 
=
 
o
u
t
p
u
t
_
d
i
r
 
/
 
w
a
v
.
n
a
m
e

 
 
 
 
s
h
u
t
i
l
.
c
o
p
y
(
w
a
v
,
 
d
e
s
t
)

 
 
 
 
s
t
e
m
_
f
i
l
e
s
[
w
a
v
.
s
t
e
m
]
 
=
 
s
t
r
(
d
e
s
t
)

 
 
 
 
p
r
i
n
t
(
f
'
C
o
p
i
e
d
 
s
t
e
m
:
 
{
w
a
v
.
s
t
e
m
}
'
)


#
 
W
r
i
t
e
 
l
y
r
i
c
s

i
f
 
l
y
r
i
c
s
_
r
e
s
u
l
t
:

 
 
 
 
(
o
u
t
p
u
t
_
d
i
r
 
/
 
'
l
y
r
i
c
s
.
j
s
o
n
'
)
.
w
r
i
t
e
_
t
e
x
t
(
j
s
o
n
.
d
u
m
p
s
(
l
y
r
i
c
s
_
r
e
s
u
l
t
)
)

 
 
 
 
p
r
i
n
t
(
'
W
r
o
t
e
 
l
y
r
i
c
s
.
j
s
o
n
'
)


#
 
W
r
i
t
e
 
b
e
a
t
s

i
f
 
b
e
a
t
s
_
r
e
s
u
l
t
:

 
 
 
 
(
o
u
t
p
u
t
_
d
i
r
 
/
 
'
b
e
a
t
s
.
j
s
o
n
'
)
.
w
r
i
t
e
_
t
e
x
t
(
j
s
o
n
.
d
u
m
p
s
(
b
e
a
t
s
_
r
e
s
u
l
t
)
)

 
 
 
 
p
r
i
n
t
(
f
'
W
r
o
t
e
 
b
e
a
t
s
.
j
s
o
n
 
(
B
P
M
:
 
{
b
e
a
t
s
_
r
e
s
u
l
t
[
"
b
p
m
"
]
}
)
'
)


#
 
W
r
i
t
e
 
k
e
y

i
f
 
k
e
y
_
r
e
s
u
l
t
:

 
 
 
 
(
o
u
t
p
u
t
_
d
i
r
 
/
 
'
k
e
y
.
j
s
o
n
'
)
.
w
r
i
t
e
_
t
e
x
t
(
j
s
o
n
.
d
u
m
p
s
(
k
e
y
_
r
e
s
u
l
t
)
)

 
 
 
 
p
r
i
n
t
(
f
'
W
r
o
t
e
 
k
e
y
.
j
s
o
n
 
(
{
k
e
y
_
r
e
s
u
l
t
[
"
k
e
y
"
]
}
)
'
)


#
 
W
r
i
t
e
 
n
o
t
e
s

n
o
t
e
s
_
a
v
a
i
l
a
b
l
e
 
=
 
[
]

f
o
r
 
i
n
s
t
r
u
m
e
n
t
,
 
t
i
m
e
l
i
n
e
 
i
n
 
n
o
t
e
s
_
b
y
_
s
t
e
m
.
i
t
e
m
s
(
)
:

 
 
 
 
(
o
u
t
p
u
t
_
d
i
r
 
/
 
f
'
n
o
t
e
s
_
{
i
n
s
t
r
u
m
e
n
t
}
.
j
s
o
n
'
)
.
w
r
i
t
e
_
t
e
x
t
(
j
s
o
n
.
d
u
m
p
s
(
t
i
m
e
l
i
n
e
)
)

 
 
 
 
n
o
t
e
s
_
a
v
a
i
l
a
b
l
e
.
a
p
p
e
n
d
(
i
n
s
t
r
u
m
e
n
t
)

 
 
 
 
p
r
i
n
t
(
f
'
W
r
o
t
e
 
n
o
t
e
s
_
{
i
n
s
t
r
u
m
e
n
t
}
.
j
s
o
n
 
(
{
l
e
n
(
t
i
m
e
l
i
n
e
)
}
 
s
e
g
m
e
n
t
s
)
'
)


#
 
C
o
m
p
u
t
e
 
s
t
e
m
 
p
r
e
s
e
n
c
e
 
(
r
e
l
a
t
i
v
e
 
R
M
S
 
e
n
e
r
g
y
 
p
e
r
 
s
t
e
m
)
 
—
 
s
h
o
w
n
 
i
n
 
t
h
e
 
U
I

i
m
p
o
r
t
 
s
o
u
n
d
f
i
l
e
 
a
s
 
s
f

i
m
p
o
r
t
 
n
u
m
p
y
 
a
s
 
n
p


s
t
e
m
_
p
r
e
s
e
n
c
e
 
=
 
{
}

r
m
s
_
v
a
l
u
e
s
 
=
 
{
}

f
o
r
 
w
a
v
_
d
e
s
t
 
i
n
 
o
u
t
p
u
t
_
d
i
r
.
g
l
o
b
(
'
*
.
w
a
v
'
)
:

 
 
 
 
s
n
a
m
e
 
=
 
w
a
v
_
d
e
s
t
.
s
t
e
m

 
 
 
 
t
r
y
:

 
 
 
 
 
 
 
 
d
a
t
a
,
 
s
r
 
=
 
s
f
.
r
e
a
d
(
s
t
r
(
w
a
v
_
d
e
s
t
)
)

 
 
 
 
 
 
 
 
r
m
s
 
=
 
f
l
o
a
t
(
n
p
.
s
q
r
t
(
n
p
.
m
e
a
n
(
d
a
t
a
 
*
*
 
2
)
)
)

 
 
 
 
 
 
 
 
r
m
s
_
v
a
l
u
e
s
[
s
n
a
m
e
]
 
=
 
r
m
s

 
 
 
 
e
x
c
e
p
t
 
E
x
c
e
p
t
i
o
n
:

 
 
 
 
 
 
 
 
p
a
s
s


i
f
 
r
m
s
_
v
a
l
u
e
s
:

 
 
 
 
m
a
x
_
r
m
s
 
=
 
m
a
x
(
r
m
s
_
v
a
l
u
e
s
.
v
a
l
u
e
s
(
)
)
 
o
r
 
1
.
0

 
 
 
 
s
t
e
m
_
p
r
e
s
e
n
c
e
 
=
 
{
k
:
 
r
o
u
n
d
(
v
 
/
 
m
a
x
_
r
m
s
 
*
 
1
0
0
,
 
1
)
 
f
o
r
 
k
,
 
v
 
i
n
 
r
m
s
_
v
a
l
u
e
s
.
i
t
e
m
s
(
)
}


#
 
W
r
i
t
e
 
m
a
n
i
f
e
s
t
 
—
 
t
h
i
s
 
i
s
 
w
h
a
t
 
t
h
e
 
a
p
p
 
r
e
a
d
s
 
t
o
 
k
n
o
w
 
w
h
a
t
'
s
 
a
v
a
i
l
a
b
l
e

m
a
n
i
f
e
s
t
 
=
 
{

 
 
 
 
'
s
o
n
g
_
i
d
'
:
 
s
o
n
g
_
i
d
,

 
 
 
 
'
t
i
t
l
e
'
:
 
i
n
p
u
t
_
p
a
t
h
.
s
t
e
m
.
r
e
p
l
a
c
e
(
'
_
'
,
 
'
 
'
)
,

 
 
 
 
'
s
t
e
m
s
'
:
 
l
i
s
t
(
s
t
e
m
_
f
i
l
e
s
.
k
e
y
s
(
)
)
,

 
 
 
 
'
n
o
t
e
s
_
a
v
a
i
l
a
b
l
e
'
:
 
n
o
t
e
s
_
a
v
a
i
l
a
b
l
e
,

 
 
 
 
'
h
a
s
_
l
y
r
i
c
s
'
:
 
l
y
r
i
c
s
_
r
e
s
u
l
t
 
i
s
 
n
o
t
 
N
o
n
e
,

 
 
 
 
'
h
a
s
_
b
e
a
t
s
'
:
 
b
e
a
t
s
_
r
e
s
u
l
t
 
i
s
 
n
o
t
 
N
o
n
e
,

 
 
 
 
'
h
a
s
_
k
e
y
'
:
 
k
e
y
_
r
e
s
u
l
t
 
i
s
 
n
o
t
 
N
o
n
e
,

 
 
 
 
'
b
p
m
'
:
 
b
e
a
t
s
_
r
e
s
u
l
t
[
'
b
p
m
'
]
 
i
f
 
b
e
a
t
s
_
r
e
s
u
l
t
 
e
l
s
e
 
N
o
n
e
,

 
 
 
 
'
k
e
y
'
:
 
k
e
y
_
r
e
s
u
l
t
[
'
k
e
y
'
]
 
i
f
 
k
e
y
_
r
e
s
u
l
t
 
e
l
s
e
 
N
o
n
e
,

 
 
 
 
'
d
e
m
u
c
s
_
m
o
d
e
l
'
:
 
S
E
P
A
R
A
T
I
O
N
_
M
O
D
E
L
,
 
 
 
#
 
l
e
g
a
c
y
 
k
e
y
 
—
 
k
e
p
t
 
f
o
r
 
a
p
p
 
c
o
m
p
a
t
i
b
i
l
i
t
y

 
 
 
 
'
s
e
p
a
r
a
t
i
o
n
_
m
o
d
e
l
'
:
 
S
E
P
A
R
A
T
I
O
N
_
M
O
D
E
L
,

 
 
 
 
'
s
e
p
a
r
a
t
i
o
n
_
e
n
g
i
n
e
'
:
 
S
E
P
A
R
A
T
I
O
N
_
E
N
G
I
N
E
,

 
 
 
 
'
s
t
e
m
_
p
r
e
s
e
n
c
e
'
:
 
s
t
e
m
_
p
r
e
s
e
n
c
e
,

}

(
o
u
t
p
u
t
_
d
i
r
 
/
 
'
m
a
n
i
f
e
s
t
.
j
s
o
n
'
)
.
w
r
i
t
e
_
t
e
x
t
(
j
s
o
n
.
d
u
m
p
s
(
m
a
n
i
f
e
s
t
,
 
i
n
d
e
n
t
=
2
)
)

p
r
i
n
t
(
f
'
\
n
M
a
n
i
f
e
s
t
 
w
r
i
t
t
e
n
:
'
)

p
r
i
n
t
(
j
s
o
n
.
d
u
m
p
s
(
m
a
n
i
f
e
s
t
,
 
i
n
d
e
n
t
=
2
)
)


#
 
─
─
 
S
a
v
e
 
t
o
 
G
o
o
g
l
e
 
D
r
i
v
e
 
(
b
y
p
a
s
s
e
s
 
b
r
o
w
s
e
r
-
d
o
w
n
l
o
a
d
 
t
i
m
e
o
u
t
)
 
─
─
─
─
─
─
─
─
─
─
─
─
─
─
─
─
─
─

#
 
D
r
i
v
e
 
i
s
 
a
l
r
e
a
d
y
 
m
o
u
n
t
e
d
 
a
t
 
/
c
o
n
t
e
n
t
/
d
r
i
v
e
 
f
r
o
m
 
C
e
l
l
 
2
.

D
R
I
V
E
_
O
U
T
P
U
T
_
D
I
R
 
=
 
P
a
t
h
(
f
'
/
c
o
n
t
e
n
t
/
d
r
i
v
e
/
M
y
D
r
i
v
e
/
{
D
R
I
V
E
_
O
U
T
P
U
T
_
F
O
L
D
E
R
}
'
)

D
R
I
V
E
_
O
U
T
P
U
T
_
D
I
R
.
m
k
d
i
r
(
p
a
r
e
n
t
s
=
T
r
u
e
,
 
e
x
i
s
t
_
o
k
=
T
r
u
e
)


p
r
i
n
t
(
'
\
n
C
r
e
a
t
i
n
g
 
z
i
p
.
.
.
'
)

z
i
p
_
p
a
t
h
 
=
 
s
h
u
t
i
l
.
m
a
k
e
_
a
r
c
h
i
v
e
(
s
o
n
g
_
i
d
,
 
'
z
i
p
'
,
 
r
o
o
t
_
d
i
r
=
'
o
u
t
p
u
t
'
,
 
b
a
s
e
_
d
i
r
=
s
o
n
g
_
i
d
)

p
r
i
n
t
(
f
'
Z
i
p
 
c
r
e
a
t
e
d
:
 
{
z
i
p
_
p
a
t
h
}
 
(
{
P
a
t
h
(
z
i
p
_
p
a
t
h
)
.
s
t
a
t
(
)
.
s
t
_
s
i
z
e
 
/
 
1
e
6
:
.
1
f
}
 
M
B
)
'
)


d
r
i
v
e
_
z
i
p
_
d
e
s
t
 
=
 
D
R
I
V
E
_
O
U
T
P
U
T
_
D
I
R
 
/
 
f
'
{
s
o
n
g
_
i
d
}
.
z
i
p
'

s
h
u
t
i
l
.
c
o
p
y
(
z
i
p
_
p
a
t
h
,
 
d
r
i
v
e
_
z
i
p
_
d
e
s
t
)

p
r
i
n
t
(
f
'
\
n
✓
 
S
a
v
e
d
 
t
o
 
G
o
o
g
l
e
 
D
r
i
v
e
:
 
M
y
 
D
r
i
v
e
/
{
D
R
I
V
E
_
O
U
T
P
U
T
_
F
O
L
D
E
R
}
/
{
s
o
n
g
_
i
d
}
.
z
i
p
'
)

p
r
i
n
t
(
'
O
p
e
n
 
G
o
o
g
l
e
 
D
r
i
v
e
 
i
n
 
y
o
u
r
 
b
r
o
w
s
e
r
 
a
n
d
 
d
o
w
n
l
o
a
d
 
t
h
e
 
z
i
p
 
a
t
 
y
o
u
r
 
c
o
n
v
e
n
i
e
n
c
e
.
'
)

p
r
i
n
t
(
f
'
E
x
t
r
a
c
t
 
s
o
 
t
h
a
t
 
b
a
c
k
e
n
d
/
d
a
t
a
/
{
s
o
n
g
_
i
d
}
/
m
a
n
i
f
e
s
t
.
j
s
o
n
 
e
x
i
s
t
s
 
(
n
o
 
e
x
t
r
a
 
n
e
s
t
i
n
g
)
.
'
)